### PR TITLE
Add `thiserror` for better display error formatting

### LINKED ISSUE #555: Error formatting is poor
Right now BDK just uses the output of `Debug` to create the display implementation. There should be a nice informative description for each error. See [`thiserror`](https://docs.rs/thiserror/latest/thiserror/).

### PR DESCRIPTION
**Pull request by yukibtc**
### Description

Resolve #555 

#### All Submissions:

* [X] I've signed all my commits
* [X] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [X] I ran `cargo fmt` and `cargo clippy` before committing

#### Bugfixes:

* [ ] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [X] I'm linking the issue being fixed by this PR

### CODE REVIEW

File: src/blockchain/compact_filters/mod.rs [hunk 1]
```diff
@@ -534,54 +533,62 @@ impl ConfigurableBlockchain for CompactFiltersBlockchain {
...
+    #[error("Peer doesn't advertise the BLOOM service flag")]
     PeerBloomDisabled,
 
     /// No peers have been specified
+    #[error("No peers have been specified")]
     NoPeers,
 
     /// Internal database error
+    #[error("Internal database error: {0}")]
     Db(rocksdb::Error),
```
Comment (rajarshimaitra) [inline src/blockchain/compact_filters/mod.rs]: As we are using `thiserror` the from impls can also be automated using `Variant(#[from] Error_type)` syntax.. 

Same fro other conversions..

**Review by rajarshimaitra — CHANGES_REQUESTED** [review]
Concept ACK..

One extra change for from impls as we are using `thiserror` so lets use its full power..

**Review by rajarshimaitra — COMMENTED** [review]
Review ACK.. Changes looks good to me.. Would let others comment with their take on using `thiserror` as a dep..

**Comment by notmandatory** [comment]
I'm in favor of this approach, but since it's adding a new dependency I think we should discuss also at our next bdk dev call, I've added it to the agenda.

**Comment by notmandatory** [comment]
@afilini in our call today we discussed making `thiserror` an optional dependency, but on second thought I don't think there's a reason to make it optional since it [only adds macros](https://docs.rs/thiserror/latest/thiserror/#details) to generate nice looking `std::error::Error` implementations, so doesn't change the bdk lib API and has no requirement to also use `anyhow`.

**Comment by danielabrozzoni** [comment]
> I don't think there's a reason to make it optional since it [only adds macros](https://docs.rs/thiserror/latest/thiserror/#details) to generate nice looking std::error::Error implementations

The reason for making it optional is that some projects want to cut dependencies to a minimum, and wouldn't use bdk if we include too many.

I agree with @afilini that it would be nice to have `thiserror` but behind a feature enabled by default

**Comment by yukibtc** [comment]
Add `thiserror` as optional dependency, in my opinion, make the code more messed up, less maintainable.
If you prefer to have fewer dependencies, it's better to avoid to add `thiserror` also as optional dependency and continue in the way of PR #814.
The result of PR #813 and #814 for the dev that use BDK library is the same, but for you the work it's doubled when you need to make some changes to error enums (using the approach of `thiserror` as optional feature).
Keep both approach it's a bad idea, in my opinion.

**Comment by notmandatory** [comment]
> Add `thiserror` as optional dependency, in my opinion, make the code more messed up, less maintainable. If you prefer to have fewer dependencies, it's better to avoid to add `thiserror` also as optional dependency and continue in the way of PR #814. The result of PR #813 and #814 for the dev that use BDK library is the same, but for you the work it's doubled when you need to make some changes to error enums (using the approach of `thiserror` as optional feature). Keep both approach it's a bad idea, in my opinion.

I agree if we want to avoid making `thiserror` a required dependency it's better to just go with your approach in #814. 

Reviewers, please check out #814, I've targeted it for the 0.27 release.

**Pull request closed without merging**